data/transactions/logic/crypto.go

```diff
@@ -30,8 +30,69 @@ import (
 	"github.com/algorand/go-algorand/protocol"
 	"github.com/algorand/go-sumhash"
 	"golang.org/x/crypto/sha3"
+
+	bls12_381fr "github.com/consensys/gnark-crypto/ecc/bls12-381/fr"
+	bls12_381mimc "github.com/consensys/gnark-crypto/ecc/bls12-381/fr/mimc"
+	bn254fr "github.com/consensys/gnark-crypto/ecc/bn254/fr"
+	bn254mimc "github.com/consensys/gnark-crypto/ecc/bn254/fr/mimc"
 )
 
+// mimc on curve BLS12-381 is implemented for compatibility with zk circuits,
+// matching the implementation in circuits generated by gnark
+func opMimcBLS12381(cx *EvalContext) error {
+	last := len(cx.Stack) - 1
+	data := cx.Stack[last].Bytes
+	if len(data)%32 != 0 {
+		return fmt.Errorf("the input data must be a multiple of 32 bytes")
+	}
+	hasher := bls12_381mimc.NewMiMC()
+	blockSize := hasher.BlockSize()
+	for i := 0; i < len(data); i += 32 {
+		d := data[i : i+32]
+		// ensure the input to hasher.Write is smaller than the curve modulus,
+		// otherwise hasher.Write will fail
+		n := new(big.Int).SetBytes(d)
+		n.Mod(n, bls12_381fr.Modulus())
+		d = n.Bytes()
+		if len(d) < blockSize {
+			d = n.FillBytes(make([]byte, blockSize))
+		}
+		hasher.Write(d)
+	}
+	hv := make([]byte, 0, hasher.Size())
+	hv = hasher.Sum(hv)
+	cx.Stack[last].Bytes = hv
+	return nil
```

Comment:
If `bls12_381mimc.NewMiMC()` is returning a `hash.Hash`, I don't understand why we can't replace the whole thing this way.  A `hash.Hash` should allow any number of bytes to be written.
```suggestion
	hasher := bls12_381mimc.NewMiMC()
	hasher.Write(cx.Stack[last].Bytes)
	hv := make([]byte, 0, hasher.Size())
	hv = hasher.Sum(hv)
	cx.Stack[last].Bytes = hv
	return nil
```
It doesn't feel like we (Algorand) should be deciding how to fold 32 byte chunks together.  Surely the `bls12_381mimc` should be doing that.

Reply:
I think I see why you did what you did - so that you could hash arbitrary bytes (including those in which a 32 byte sequence happens to be larger than the modulus).  I suppose that makes sense, but it freaks me out that this package is exporting a `hash.Hash` that does not take arbitrary data.
It also freaks me out that we now have a hash function that has _known_ collisions.  Imagine any input that has a 32 byte sequence that is interpretted as 1 higher than the modulus.  It has the same hash as if you replace that sequence with 32 bytes in which nly the last byte is set, to 1.

Reply:
`fr.WriteString()` is available for arbitrary inputs, so we could feed `cx.Stack[last].Bytes` to it instead of `Write()`.

However, that method is defined on the (unexported) `mimc.digest` type, while `NewMiMC` returns a `hash.Hash` which does not have that method declared. So I think it's uncallable, though that is surely not intended or it would be called `writeString`.

Reply:
In my opinion, the definition of `Write()` should be considered a bug in gnark.  The `hash.Hash` interface says:
```
	// Write (via the embedded io.Writer interface) adds more data to the running hash.
	// It never returns an error.
```

Reply:
I've submitted this.  I don't know what they'll say, because I don't know the state of standardization of mimc hashing.
https://github.com/Consensys/gnark-crypto/issues/504